tests/framework/testutils/execute.go

```diff
@@ -23,15 +24,27 @@ import (
 )
 
 func ExecuteWithTimeout(t *testing.T, timeout time.Duration, f func()) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	ExecuteUntil(ctx, t, f)
+}
+
+func ExecuteUntil(ctx context.Context, t *testing.T, f func()) {
+	deadline, deadlineSet := ctx.Deadline()
 	donec := make(chan struct{})
 	go func() {
 		defer close(donec)
 		f()
 	}()
 
 	select {
-	case <-time.After(timeout):
-		testutil.FatalStack(t, fmt.Sprintf("test timed out after %v", timeout))
+	case <-ctx.Done():
+		msg := ctx.Err().Error()
+		if deadlineSet {
+			timeout := time.Until(deadline)
```

Comment:
This might not be correct. The `timeout` here isn't the original 10 seconds any more. For simplicity, you can just remove it for now.

Reply:
It will write `9m59s` or similar, but I think it's still correct. Only issue is that it will not be easy to find the source of timeout, however this is expected when we introduce hierarchical timeouts.

Reply:
Ups, wanted to refactor but forgot about `time.Until` needs to be called at the top.